Add FaaS metrics semantic conventions

skonto opened the pull request:
Fixes #1013

## Changes
Adds metrics and metric labels for functions
<!-- Please provide a brief description of the changes here. Update the
`CHANGELOG.md` for non-trivial changes. If `CHANGELOG.md` is updated,
also be sure to update `spec-compliance-matrix.md` if necessary.
-->

I am proposing also changing the `faas` prefix to `func` as not all environments are of that kind.
Do I need to discuss that change in a WG meeting?

/cc @kolanos @kenfinnigan @jmacd

arminru commented:
@skonto The namespace/prefix is shared with the trace and resource semantic conventions:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/faas.md
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/faas.md

If you think changing it would make sense for all three of them, please open a separate issue to discuss this.

arminru commented on specification/metrics/semantic_conventions/faas-metrics.md:
```diff
@@ -0,0 +1,49 @@
+# General
```
Please add this new file to the list in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md and also add an entry in https://github.com/open-telemetry/opentelemetry-specification/blob/main/CHANGELOG.md. Thanks!

skonto replied:
> If you think changing it would make sense for all three of them, please open a separate issue to discuss this.
I saw that that is why I set this PR to wip, wanted see if I can do it here, but I agree better to start a discussion first.

skonto commented:
@arminru 
> If you think changing it would make sense for all three of them, please open a separate issue to discuss this.

I saw that that is why I set this PR to wip, wanted see if I can do it here, but I agree better to start a discussion first.

kenfinnigan commented:
Question for maintainers. What are the expectations around duplication of data, or preferences for where it resides?

Taking a specific example, the `faas.invoked_name`, etc labels can be found if a `Resource` with that information is set, correct? Should the specification be preferring the information being set in a `Resource` and that resource is then attached to the metric, vs labels with identical information being added?

skonto commented:
@arminru gentle ping for review.

arminru commented on specification/metrics/semantic_conventions/faas-metrics.md:
```diff
@@ -0,0 +1,57 @@
+# General
+
```
```suggestion

**Status**: [Experimental](../../document-status.md)

```

arminru commented:
> Question for maintainers. What are the expectations around duplication of data, or preferences for where it resides?
> 
> Taking a specific example, the `faas.invoked_name`, etc labels can be found if a `Resource` with that information is set, correct? Should the specification be preferring the information being set in a `Resource` and that resource is then attached to the metric, vs labels with identical information being added?

@kenfinnigan
Only the following FaaS attributes are defined for the resource: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/faas.md
There is no overlap with the metrics attributes added by this PR as far as I can see.

arminru commented:
@open-telemetry/specs-metrics-approvers @reyang @jmacd Could you please take a look?

kenfinnigan commented:
> > Question for maintainers. What are the expectations around duplication of data, or preferences for where it resides?
> > Taking a specific example, the `faas.invoked_name`, etc labels can be found if a `Resource` with that information is set, correct? Should the specification be preferring the information being set in a `Resource` and that resource is then attached to the metric, vs labels with identical information being added?
> 
> @kenfinnigan
> Only the following FaaS attributes are defined for the resource: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/faas.md
> There is no overlap with the metrics attributes added by this PR as far as I can see.

Would `faas.invoked_name` here be the same as `faas.name` on the Resource?

I also wasn't sure whether there would be overlap between `faas.invoked_provider` here, and `cloud.provider` on the cloud resource: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/cloud.md

arminru commented:
> Would `faas.invoked_name` here be the same as `faas.name` on the Resource?
> 
> I also wasn't sure whether there would be overlap between `faas.invoked_provider` here, and `cloud.provider` on the cloud resource: [https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/cloud.md](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/cloud.md?rgh-link-date=2021-06-09T12%3A53%3A58Z)

@kenfinnigan In the [trace semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/faas.md) things are quite clear since incoming and outgoing invocations are distinguished clearly. In tracing, `faas.invoked_name` on the *client* span which is invoking the function would be the same as the `faas.name` resource attribute on the function being invoked.
I would think that this convention here would follow the same semantics, so a `faas.invoked_name` label and `faas.name` resource attribute would only be present on the same process if it is a function itself (`faas.name`) making an outgoing call to yet another function (`faas.invoked_name`). It would, however, only have the same value if it would call itself recursively.
Likewise for `faas.invoked_provider` and `cloud.provider`. A process running in Google cloud would have `cloud.provider=gcp` set on its resource and if it had a metric tracking outgoing invocations towards a Lambda function would the label `faas.invoked_provider=aws` set, for example.

@skonto I think it would make sense to use the same structure as in the trace semantic conventions (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/faas.md) here as well to make this clearer.

kenfinnigan commented:
@arminru Thanks for the detail. The piece I was missing is that metrics could have an equivalent "in process" and "outgoing" as traces, as I'd presumed metrics would only be dealing with the current process only

skonto commented:
@arminru Yes I can do that

@kenfinnigan 

> as I'd presumed metrics would only be dealing with the current process only

I thought the same but then it might be the case that we have a metric that counts invocations from within a process
and may tag the metric with that label explicitly.

cijothomas commented on specification/metrics/semantic_conventions/faas-metrics.md:
```diff
@@ -0,0 +1,57 @@
+
+* [AWS Lambda Metrics](https://docs.aws.amazon.com/lambda/latest/dg/monitoring-metrics.html)
+* [AWS Lambda Insight Metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Lambda-Insights-metrics.html)
+* [Azure Functions Metrics](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/metrics-supported)
```
```suggestion
* [Azure Functions Metrics](https://docs.microsoft.com/azure/azure-monitor/platform/metrics-supported)
```

skonto commented:
@arminru I updated the spec, I dont want to repeat the info at the trace side (added a link), however I changed the structure according to other metrics specs.

arminru commented on specification/metrics/semantic_conventions/faas-metrics.md:
```diff
@@ -0,0 +1,74 @@
+| `faas.invoked_provider` | Yes | Cloud provider of the invoked function. Corresponds to the resource `cloud.provider`. Example: `aws` |
+| `faas.invoked_region` | Yes | Cloud provider region of invoked function. Corresponds to resource `cloud.region`. Example: `us-east-1` |
+
+More details on these labels and the function name and the difference compared to the faas.invoked_name can be found at the related [faas tracing specification](../../trace/semantic_conventions/faas.md).
```
```suggestion
More details on these labels, the function name and the difference compared to the faas.invoked_name can be found at the related [FaaS tracing specification](../../trace/semantic_conventions/faas.md).
```

skonto commented on specification/metrics/semantic_conventions/faas-metrics.md:
```diff
@@ -0,0 +1,74 @@
+
+## Labels
+
+Below is a table of the labels that SHOULD be included on FaaS metric events.
```
Ok I will make a clarification there.

arminru replied:
I think the SHOULD for the others was actually fine. Now the spec is unclear of what is excepted here.

skonto commented on specification/metrics/semantic_conventions/faas-metrics.md:
```diff
@@ -0,0 +1,74 @@
+| `faas.invoked_name` | Yes | Name of the invoked function. Example: `my-function` |
+| `faas.invoked_provider` | Yes | Cloud provider of the invoked function. Corresponds to the resource `cloud.provider`. Example: `aws` |
+| `faas.invoked_region` | Yes | Cloud provider region of invoked function. Corresponds to resource `cloud.region`. Example: `us-east-1` |
+
```
Ok I can add that one.

skonto commented:
@arminru ready for another round.

arminru commented:
@skonto seems like this was left behind while I was on vacation, sorry about that. LGTM in general but it would be great if one of our @open-telemetry/specs-metrics-approvers could have a look as well.

Please make sure to use the current instrument names as described in #1788 and to incorporate the changes introduced by #1781.

skonto commented:
Ok I will update this.

skonto commented:
@arminru I updated the instrument types, regarding the other update you mention, I suspect the resource labels can be mixed in with the ones in this PR if they have to, I am not sure if I need to reference them though at all.

arminru commented on CHANGELOG.md:
```diff
@@ -109,6 +109,7 @@ release.
   [#1731](https://github.com/open-telemetry/opentelemetry-specification/pull/1731),
   [#1733](https://github.com/open-telemetry/opentelemetry-specification/pull/1733))
 - Mark relevant portions of Metrics Data Model stable ([#1728](https://github.com/open-telemetry/opentelemetry-specification/pull/1728))
+- Add FaaS metrics semantic conventions ([#1736](https://github.com/open-telemetry/opentelemetry-specification/pull/1736))
```
Seems like this slipped down when merging/rebasing, please move it up to the `Unreleased` section.

arminru commented on specification/metrics/semantic_conventions/faas-metrics.md:
```diff
@@ -0,0 +1,77 @@
+
+The conventions described in this section are FaaS (function as a service) specific. When FaaS operations occur,
+metric events about those operations will be generated and reported to provide insights into the
+operations. By adding FaaS labels to metric events it allows for finely tuned filtering.
+
+**Disclaimer:** These are initial FaaS metric instruments and labels but more may be added in the future.
+
+<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->
+
+<!-- toc -->
+- [Metric Instruments](#metric-instruments)
+  * [FaaS Invocations](#faas-invocations)
+- [Labels](#labels)
```
```suggestion
operations. By adding FaaS attributes to metric events it allows for finely tuned filtering.

**Disclaimer:** These are initial FaaS metric instruments and attributes but more may be added in the future.

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->
- [Metric Instruments](#metric-instruments)
  * [FaaS Invocations](#faas-invocations)
- [Attributes ](#attributes)
```

Metric labels were renamed to attributes (consolidated with trace and resource) when the Metrics API was reworked.

arminru commented on specification/metrics/semantic_conventions/faas-metrics.md:
```diff
@@ -0,0 +1,77 @@
+| `faas.cpu_usage` | Histogram | milliseconds | Distribution of cpu usage per invocation |
+| `faas.net_io` | Histogram | bytes | Distribution of net I/O usage per invocation |
+
+## Labels
+
+Below is a table of the labels to be included on FaaS metric events.
+
+| Name | Recommended | Notes and examples |
+|------|-------------|--------------------|
+| `faas.trigger` | Yes | Type of the trigger on which the function is invoked. SHOULD be one of: `datasource`, `http`, `pubsub`, `timer`, `other` |
+| `faas.invoked_name` | Yes | Name of the invoked function. Example: `my-function` |
+| `faas.invoked_provider` | Yes | Cloud provider of the invoked function. Corresponds to the resource `cloud.provider`. Example: `aws` |
+| `faas.invoked_region` | Yes | Cloud provider region of invoked function. Corresponds to resource `cloud.region`. Example: `us-east-1` |
+
+More details on these labels, the function name and the difference compared to the faas.invoked_name can be found at the related [FaaS tracing specification](../../trace/semantic_conventions/faas.md).
```
```suggestion
## Attributes

Below is a table of the attributes to be included on FaaS metric events.

| Name | Recommended | Notes and examples |
|------|-------------|--------------------|
| `faas.trigger` | Yes | Type of the trigger on which the function is invoked. SHOULD be one of: `datasource`, `http`, `pubsub`, `timer`, `other` |
| `faas.invoked_name` | Yes | Name of the invoked function. Example: `my-function` |
| `faas.invoked_provider` | Yes | Cloud provider of the invoked function. Corresponds to the resource `cloud.provider`. Example: `aws` |
| `faas.invoked_region` | Yes | Cloud provider region of invoked function. Corresponds to resource `cloud.region`. Example: `us-east-1` |

More details on these attributes, the function name and the difference compared to the faas.invoked_name can be found at the related [FaaS tracing specification](../../trace/semantic_conventions/faas.md).
```

arminru reviewed (commented):
@skonto Thanks! Seems like the PR I referred to only added to the FaaS trace semconv but did not touch the attributes which you took over, so no need for updating then. The resource attributes don't have to recited here, yes.

@open-telemetry/specs-metrics-approvers Please take a look.

arminru reviewed (approved):
Thanks, @skonto!

carlosalberto commented:
@open-telemetry/specs-metrics-approvers Please take a look and review/approve.

skonto commented:
@jmacd gentle ping, can I get merge pls?
